harness/determined/core/_metrics.py

```diff
@@ -0,0 +1,168 @@
+import logging
+import queue
+import threading
+from typing import Any, Dict, List, Optional
+
+from determined.common import api
+from determined.common.api import bindings
+
+logger = logging.getLogger("determined.core")
+
+
+class MetricsContext:
+    """Gives access to metrics reporting during trial tasks.
+
+    Metrics reported to ``MetricsContext`` are published to a queue, which is consumed by a
+    background thread that reports them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        trial_id: int,
+        run_id: int,
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._error_queue = queue.Queue()
+        self._shipper = _Shipper(
+            session=self._session,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            error_queue=self._error_queue,
+        )
+
+    def report(
+        self,
+        group: str,
+        steps_completed: int,
+        metrics: Dict[str, Any],
+        batch_metrics: Optional[List[Dict[str, Any]]] = None,
+    ) -> None:
+        # Check for thread exceptions here since we're not polling.
+        if not self._error_queue.empty():
+            err_msg = self._error_queue.get(block=False)
+            logger.error(f"Error reporting metrics: {err_msg}")
+            raise err_msg
+
```

Comment:
should we also check/flush the error queue on close? 

e.g. what if user
1. logs only one metric
2. it fails in the background
3. user logs no more and exits

it'd be good to wait for that report to flush & log error, if any.

Reply:
good idea. done.